Code/Tools/AssetProcessor/Platform/Linux/native/FileWatcher/FileWatcher_linux.cpp

```diff
@@ -34,38 +71,71 @@ bool FileWatcher::PlatformImplementation::Initialize()
     {
         return;
     }
+      
+    ::close(m_inotifyHandle);
+    m_inotifyHandle = -1;
 
+    // signal the thread to awaken in case its blocked on read.
+    AZ::u64 flagValue = 1;
+    if (write(m_wakeThreadHandle, &flagValue, sizeof(flagValue)) != sizeof(flagValue))
     {
-        QMutexLocker lock{&m_handleToFolderMapLock};
-        for (const auto& watchHandle : m_handleToFolderMap.keys())
-        {
-            inotify_rm_watch(m_inotifyHandle, watchHandle);
-        }
-        m_handleToFolderMap.clear();
+        AZ_ErrorOnce("FileWatcher", false, "Was unable to write to the wake-up-the-thread handle, this may cause timeouts and deadlocks.\n");
     }
+}
 
-    ::close(m_inotifyHandle);
-    m_inotifyHandle = -1;
+void FileWatcher::PlatformImplementation::Finalize()
+{
+    // This happens after the watch thread has already joined.
+    
+    // note from the man-page of inotify " all associated watches are automatically freed." - no need to call inotify_rm_watch on each handle.
+    if (m_wakeThreadHandle >= 0)
+    {
+        close(m_wakeThreadHandle);
+        m_wakeThreadHandle = -1;
+    }
+    m_handleToFolderMap.clear();
+    m_alreadyNotifiedCreate.clear();
 }
 
-bool FileWatcher::PlatformImplementation::TryToWatch(const QString &pathStr, int* errnoPtr)
+bool FileWatcher::PlatformImplementation::TryToWatch(const QString &pathStr, int* errnoPtr, FileWatcher& source)
 {
     AZ::IO::FixedMaxPathString path(pathStr.toUtf8().constData());
+    // note:  IN_MASK_CREATE will set EEXIST if the directory already has a watch established on it.
+    // note:  IN_ONLYDIR will set ENOTDIR if the watch was attempted to establish on a file instead of a dir.
     int watchHandle = inotify_add_watch(
-        m_inotifyHandle, path.c_str(), IN_CREATE | IN_CLOSE_WRITE | IN_DELETE | IN_DELETE_SELF | IN_MODIFY | IN_MOVE);
+        m_inotifyHandle, path.c_str(), IN_CREATE | IN_CLOSE_WRITE | IN_DELETE | IN_DELETE_SELF | IN_MODIFY | IN_MOVE | IN_MOVE_SELF | IN_MASK_CREATE | IN_ONLYDIR); 
     const auto err = errno; // Only contains relevant information if we actually failed
 
     if (watchHandle < 0)
     {
+        if (err == ENOTDIR)
+        {
+            // the dir being watched was deleted and replaced by a file before we managed to watch it.
+            // this is okay, and absorbing this removes a race condition.
+            DEBUG_FILEWATCHER("Not adding an additional file watch for %s - it is not a directory s\n", pathStr.toUtf8().constData());
+            return false;
+        }
+        if (err == EEXIST)
+        {
+            // the dir already has a watch handle on it that belongs to m_inotifyHandle.
+            // avoid duplicating watches.  This is also okay to absorb but we do return true to indicate
+            // that we successfully have a watch on it.
+            DEBUG_FILEWATCHER("Not adding an additional file watch for %s - already exists\n", pathStr.toUtf8().constData());
+            return false;
```

Comment:
I'll update the comment.  The behavior is correct - that is, if a watch is already established on that folder, it will have already established watches on its subtree.